### PR TITLE
Qdtrack

### PR DESCRIPTION
**Pull request by mageofboy**

### CODE REVIEW

**Comment by ICGog** [comment]
@mageofboy Is this PR ready for review?

**Comment by mageofboy** [comment]
> @mageofboy Is this PR ready for review?

Sorry for the delayed response. Yes, it is mostly ready, outside of some reformatting

File: configs/tracking.conf [hunk 1]
```diff
@@ -15,7 +15,7 @@
 --simulator_num_people=200
 --simulator_num_vehicles=10
 ######### Other config #########
---visualize_tracked_obstacles
+# --visualize_tracked_obstacles
```
Comment (ICGog) [inline configs/tracking.conf:18]: Can you please revert changes to this config file. I think it's best if this config file uses sort by default because it's a well tested tracker that doesn't have many dependencies.

File: pylot/perception/flags.py [hunk 2]
```diff
@@ -84,6 +84,16 @@
...
                   'CenterTrack available models')
 
+# QDTrack tracking flags.
+flags.DEFINE_string(
+    'qd_track_model_path',
+    'dependencies/models/tracking/qd_track/qdtrack-frcnn_r50_fpn_12e_bdd100k-13328aed.pth', #FIX
+    'Path to the model')
+flags.DEFINE_string(
+    'qd_track_config_path',
+    'dependencies/qdtrack/configs/qdtrack-frcnn_r50_fpn_12e_bdd100k.py', #FIX
```
Comment (ICGog) [inline pylot/perception/flags.py:94]: Why is there a fix comment?

File: pylot/perception/flags.py [hunk 3]
```diff
@@ -84,6 +84,16 @@
...
+    'Path to the model')
```
Comment (ICGog) [inline pylot/perception/flags.py:95]: I guess this is the Path to the model configuration.

File: scripts/lanenet_data_postprocess.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+from absl import app, flags
```
Comment (ICGog) [inline scripts/lanenet_data_postprocess.py:1]: Can you please create a separate PR for this file because it is not related to this PR?

File: pylot/perception/tracking/qd_track_operator.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        image_np = msg.frame.as_bgr_numpy_array()
+        results = inference_model(self.model, image_np, self.frame_id)
+        self.frame_id += 1
+
+        bbox_result, track_result = results.values()
+        track_bboxes = np.zeros((0, 5))
+        track_ids = np.zeros((0), dtype=int)
+        track_labels = np.zeros((0), dtype=int)
+        obstacle_count = 0
+        for k, v in track_result.items():
```
Comment (ICGog) [inline pylot/perception/tracking/qd_track_operator.py:58]: It looks like you don't actually need the entire track_bboxes, track_ids, and track_labels arrays. So, could you please merge the two loops into a single one, and directly extract track_id, bboxes and label_id from k and v?

File: install.sh [hunk 6]
```diff
@@ -141,6 +146,15 @@ sudo apt-get install llvm-9
 export LLVM_CONFIG=/usr/bin/llvm-config-9
 python3 setup.py build develop --user
 
+###### Install QDTrack ######
+cd $PYLOT_HOME/dependencies/
+git clone https://github.com/mageofboy/qdtrack.git
+git clone https://github.com/open-mmlab/mmdetection.git
+cd mmdetection
+python3 setup.py develop #need to add mmcv
```
Comment (ICGog) [inline install.sh:154]: How reliable is mmdetection's installation? Did it work out of the box for you?

**Review by ICGog — COMMENTED** [review]
It's looking good, but please address my comments before we merge in the PR.

**Comment by ICGog** [comment]
Also, can you please make the code style changes necessary to ensure that the CI passes? You can see the suggested changes here: https://github.com/erdos-project/pylot/pull/183/checks?check_run_id=2488080643

File: pylot/perception/flags.py (hunk 3 above)
Comment (mageofboy) [reply pylot/perception/flags.py:95]: Yea, it's the qd track model path but it's currently the path to the bdd model, rather than the waymo model

File: install.sh (hunk 6 above)
Comment (mageofboy) [reply install.sh:154]: unfortunately it is not reliable and installation depends on torch and cuda version

File: configs/tracking.conf [hunk 7]
```diff
@@ -4,7 +4,7 @@
 --obstacle_detection_gpu_memory_fraction=0.4
 ######### Tracker config #########
 --obstacle_tracking
---tracker_type=sort
+--tracker_type=qd_track
```
Comment (ICGog) [inline configs/tracking.conf:7]: Can you please set the tracker_type to sort instead of qd_track?

**Review by ICGog — COMMENTED** [review]
Looks good to me! There's only one comment left to address.
